src/lib/math/numbertheory/reducer.cpp

```diff
@@ -25,15 +25,36 @@ Modular_Reducer::Modular_Reducer(const BigInt& mod) {
    m_mod_words = 0;
 
    if(mod > 0) {
-      m_modulus = mod;
-      m_mod_words = m_modulus.sig_words();
-
-      // Compute mu = floor(2^{2k} / m)
-      m_mu.set_bit(2 * BOTAN_MP_WORD_BITS * m_mod_words);
-      m_mu = ct_divide(m_mu, m_modulus);
+      *this = Modular_Reducer::for_secret_modulus(mod);
    }
 }
 
+Modular_Reducer Modular_Reducer::for_secret_modulus(const BigInt& mod) {
+   BOTAN_ARG_CHECK(!mod.is_zero(), "Modulus cannot be zero");
+   BOTAN_ARG_CHECK(!mod.is_negative(), "Modulus cannot be negative");
+
+   // Left uninitialized if mod == 0
+   size_t mod_words = mod.sig_words();
+
+   // Compute mu = floor(2^{2k} / m)
+   const size_t mu_bits = 2 * BOTAN_MP_WORD_BITS * mod_words;
+   BigInt mu = ct_divide_pow2k(mu_bits, mod);
+   return Modular_Reducer(mod, mu, mod_words);
+}
+
+Modular_Reducer Modular_Reducer::for_public_modulus(const BigInt& mod) {
+   BOTAN_ARG_CHECK(!mod.is_zero(), "Modulus cannot be zero");
+   BOTAN_ARG_CHECK(!mod.is_negative(), "Modulus cannot be negative");
+
+   // Left uninitialized if mod == 0
+   size_t mod_words = mod.sig_words();
+
+   // Compute mu = floor(2^{2k} / m)
+   const size_t mu_bits = 2 * BOTAN_MP_WORD_BITS * mod_words;
+   BigInt mu = BigInt::power_of_2(mu_bits) / mod;
+   return Modular_Reducer(mod, mu, mod_words);
+}
```

Comment:
Let's avoid allocations and copies as much as we can. Similarly for the private variant.

```suggestion
Modular_Reducer Modular_Reducer::for_public_modulus(BigInt mod) {
   BOTAN_ARG_CHECK(!mod.is_zero(), "Modulus cannot be zero");
   BOTAN_ARG_CHECK(!mod.is_negative(), "Modulus cannot be negative");

   // Left uninitialized if mod == 0
   size_t mod_words = mod.sig_words();

   // Compute mu = floor(2^{2k} / m)
   const size_t mu_bits = 2 * BOTAN_MP_WORD_BITS * mod_words;
   BigInt mu = BigInt::power_of_2(mu_bits) / mod;
   return Modular_Reducer(std::move(mod), std::move(mu), mod_words);
}
```

Reply:
Done for `mu` there is no point with the modulus imo